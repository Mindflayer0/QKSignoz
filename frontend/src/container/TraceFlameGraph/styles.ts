import styled from 'styled-components';

export const SpanItemContainer = styled.div`
	position: absolute;
	top: ${(props) => props.topOffset}px;
	left: ${(props) => props.leftOffset}%;
	width: ${(props) => props.width}%;
	height: 10px;
	margin: 1px 0;
	background-color: #0ca;
	border-radius: 5px;
	opacity: ${(props) => (props.selected ? 1 : 0.3)};
`;

export const TraceFlameGraphContainer = styled.div`
	position: relative;
	border: 1px dotted #ccc;
	width: 100%;
	height: ${({ height }) => (height ? height : 120)}px;
`;
