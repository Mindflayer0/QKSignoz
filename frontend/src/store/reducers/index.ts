import { combineReducers } from 'redux';

import appReducer from './app';
import dashboardReducer from './dashboard';
import globalTimeReducer from './global';
import { LogsReducer } from './logs';
import metricsReducers from './metric';
import { PipelineReducer } from './pipeline';
import { ServiceMapReducer } from './serviceMap';
import traceReducer from './trace';
import { usageDataReducer } from './usage';

const reducers = combineReducers({
	traces: traceReducer,
	usageDate: usageDataReducer,
	globalTime: globalTimeReducer,
	serviceMap: ServiceMapReducer,
	dashboards: dashboardReducer,
	app: appReducer,
	metrics: metricsReducers,
	logs: LogsReducer,
	pipeline: PipelineReducer,
});

export type AppState = ReturnType<typeof reducers>;

export default reducers;
