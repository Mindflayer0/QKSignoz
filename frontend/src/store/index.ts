import { applyMiddleware, compose, createStore } from 'redux';
import type { ThunkMiddleware } from 'redux-thunk';
import thunk from 'redux-thunk';

import reducers, { AppState } from './reducers';

const composeEnhancers =
	(window && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

const store = createStore(
	reducers,
	// @TODO Add Type for AppActions also
	composeEnhancers(applyMiddleware(thunk as ThunkMiddleware<AppState>)),
);

export default store;
