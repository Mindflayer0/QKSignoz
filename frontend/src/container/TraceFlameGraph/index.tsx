import React, { useState } from 'react';
import { Tooltip } from 'antd';
import { pushDStree, span } from 'store/actions';
import { SpanItemContainer, TraceFlameGraphContainer } from './styles';

/**
 * Traverses the Span Tree data and returns the relevant meta data.
 * Metadata includes globalStart, globalEnd,
 */

// const spanServiceNameToColourMapping = (spans: span[]) => {
// 	const serviceNameSet = new Set();
// 	spans.forEach((spanItem) => {
// 		serviceNameSet.add(spanItem[3]);
// 	});
// 	debugger;
// };
const getMetaDataFromSpanTree = (treeData: pushDStree) => {
	let globalStart = Number.POSITIVE_INFINITY;
	let globalEnd = Number.NEGATIVE_INFINITY;
	let totalSpans = 0;
	let levels = 1;
	const traverse = (treeNode: pushDStree, level: number = 0) => {
		if (!treeNode) {
			return;
		}
		totalSpans++;
		levels = Math.max(levels, level);
		const startTime = treeNode.startTime;
		const endTime = startTime + treeNode.value;
		globalStart = Math.min(globalStart, startTime);
		globalEnd = Math.max(globalEnd, endTime);

		for (const childNode of treeNode.children) {
			traverse(childNode, level + 1);
		}
	};
	traverse(treeData, 1);

	globalStart = globalStart * 1e6;
	globalEnd = globalEnd * 1e6;

	return {
		globalStart,
		globalEnd,
		spread: globalEnd - globalStart,
		totalSpans,
		levels,
	};
};

const SpanItem = ({
	topOffset = 0,
	leftOffset = 0,
	width = 10,
	nodeData,
	onSpanSelect,
}: {
	topOffset: number;
	leftOffset: number;
	width: number;
	nodeData: pushDStree;
	onSpanSelect: Function;
}) => {
	const [isSelected, setIsSelected] = useState<boolean>(false);
	const handleHover = () => {
		setIsSelected((prevState) => !prevState);
	};
	return (
		<>
			<Tooltip placement="top" title={nodeData.name} key={nodeData.name}>
				<SpanItemContainer
					onClick={() => onSpanSelect(nodeData)}
					topOffset={topOffset}
					leftOffset={leftOffset}
					width={width}
					onMouseEnter={() => {
						setIsSelected(true);
						console.log('Enter');
					}}
					onMouseLeave={() => {
						setIsSelected(false);
						console.log('leave');
					}}
					selected={isSelected}
				></SpanItemContainer>
			</Tooltip>
		</>
	);
};

const TraceFlameGraph = (props: { treeData: pushDStree }) => {
	if (!props.treeData || props.treeData.id === 'empty') {
		return null;
	}

	const [treeData, setTreeData] = useState<pushDStree>(props.treeData);

	const handleSpanSelection = (selectedNodeData: pushDStree) => {
		setTreeData(selectedNodeData);
	};
	const {
		globalStart,
		globalEnd,
		spread,
		totalSpans,
		levels,
	} = getMetaDataFromSpanTree(treeData);
	const vSize = 12;

	const RenderSpanRecursive = ({
		level = 0,
		nodeData,
		parentLeftOffset = 0,
	}: {
		nodeData: pushDStree;
		level?: number;
		parentLeftOffset?: number;
	}) => {
		if (!nodeData) {
			return null;
		}
		const leftOffset = ((nodeData.startTime * 1e6 - globalStart) * 1e8) / spread;

		const width = (nodeData.value * 1e8) / spread;
		return (
			<>
				<SpanItem
					topOffset={level * vSize}
					leftOffset={leftOffset}
					width={width}
					nodeData={nodeData}
					onSpanSelect={handleSpanSelection}
				/>
				{nodeData.children.map((childData) => (
					<RenderSpanRecursive
						level={level + 1}
						nodeData={childData}
						key={childData.id}
						parentLeftOffset={leftOffset + parentLeftOffset}
					/>
				))}
			</>
		);
	};
	return (
		<>
			<TraceFlameGraphContainer height={vSize * levels}>
				<RenderSpanRecursive nodeData={treeData} />
			</TraceFlameGraphContainer>
			<button onClick={() => setTreeData(props.treeData)}>Reset</button>
		</>
	);
};

export default TraceFlameGraph;
