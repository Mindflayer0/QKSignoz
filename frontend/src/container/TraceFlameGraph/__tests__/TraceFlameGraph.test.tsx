import { expect } from '@jest/globals';
import React from 'react';
import { render } from '@testing-library/react';
import TraceFlameGraph from 'container/TraceFlameGraph';
import { Provider } from 'react-redux';
import store from 'store';

it('loads and displays greeting', async () => {
	const { asFragment } = render(
		<Provider store={store}>
			<TraceFlameGraph
				{...{
					hoveredSpanId: '',
					intervalUnit: { multiplier: 0, name: 'm' },
					onSpanHover: () => {},
					onSpanSelect: () => {},
					selectedSpanId: '',
					traceMetaData: [],
					treeData: {
						children: [],
						id: '',
						name: '',
						serviceColour: '',
						serviceName: '',
						startTime: 0,
						tags: [],
						time: 0,
						value: 0,
						event: [],
						hasError: false,
						parent: undefined,
					},
					key: '',
				}}
			/>
		</Provider>,
	);
	expect(asFragment()).toMatchSnapshot();
});
