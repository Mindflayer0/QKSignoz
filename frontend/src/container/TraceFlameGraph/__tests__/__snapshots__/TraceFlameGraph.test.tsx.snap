// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loads and displays greeting 1`] = `
<DocumentFragment>
  <div
    class="sc-gsDKAQ jFDWPs"
    height="NaN"
  >
    <div
      class="sc-bdvvtL fnQkDB"
      title="
0 m"
      width="NaN"
    />
  </div>
</DocumentFragment>
`;
